x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/ScalarFunction.java

```diff
@@ -110,31 +114,48 @@ protected ScriptTemplate scriptWithScalar(ScalarFunction scalar) {
     }
 
     protected ScriptTemplate scriptWithAggregate(AggregateFunction aggregate) {
-        String template = basicTemplate(aggregate);
-        return new ScriptTemplate(processScript(template),
-                paramsBuilder().agg(aggregate).build(),
-                dataType());
+        Tuple<String, DataType> template = basicTemplate(aggregate);
+        ParamsBuilder paramsBuilder = paramsBuilder().agg(aggregate);
+        if (template.v2() != null) {
+            paramsBuilder.variable(template.v2().name());
+        }
+        return new ScriptTemplate(processScript(template.v1()), paramsBuilder.build(), dataType());
     }
 
     // This method isn't actually used at the moment, since there is no grouping function (ie HISTOGRAM)
     // that currently results in a script being generated
     protected ScriptTemplate scriptWithGrouping(GroupingFunction grouping) {
-        String template = basicTemplate(grouping);
-        return new ScriptTemplate(processScript(template),
-                paramsBuilder().grouping(grouping).build(),
-                dataType());
+        Tuple<String, DataType> template = basicTemplate(grouping);
+        if (template.v2() != null) {
+            throw new QlIllegalArgumentException("Unexpected data type [" + template.v2() + "] for grouping function [" + grouping + "]");
+        }
+        return new ScriptTemplate(processScript(template.v1()),
+            paramsBuilder().grouping(grouping).build(),
+            dataType());
     }
 
     // FIXME: this needs to be refactored to account for different datatypes in different projects (ie DATE from SQL)
-    private String basicTemplate(Function function) {
+    private Tuple<String, DataType> basicTemplate(Function function) {
```

Comment:
Returning the tuple is clunky since the initial method was designed to basic template as a string and allow postprocessing.
Instead do the numeric handling after the script creation and bind the type as a parameter to the newly created script.
Looking at the code this needs to occur only in `scriptWithAggregate` so it can be localized there and avoid over generalization.

Reply:
Adressed.